R/util_dfc.R

```diff
@@ -118,13 +118,178 @@ dfc.compare_df_values <- function(df1, df2, id_col, verbose = FALSE) {
   return(list(diff_df = dfdiff,
               row_summary = dfdiff_sum_rows,
-              col_summary = dfdiff_sum_cols))
+              col_summary = dfdiff_sum_cols,
+              side_by_side_df = side_by_side_df))
+
+}
+
+
+dfc.strip_for_comparison <- function(DF){
+  # formats a DF to eliminate frequent causes of unimportant mismatches, including
+  # differently-formatted blanks, rounding errors, and special characters
+  clean_blanks <- function(DF){
+    DF %>%
+      util.apply_columns(function(x) ifelse(util.is_blank(x), NA, x)) %>%
+      util.apply_columns(function(x) if(all(is.na(x))) return(rep(NA, length(x))) else return(x))
+  }
+  
+  DF %>%
+    clean_blanks %>%
+    util.apply_columns(., util.as_numeric_if_number) %>%
+    util.round_df(., 5) %>%
+    util.apply_columns(util.strip_non_ascii) %>%
+    util.apply_columns(util.strip_special_characters)
+}
 
+compare_to_previous <- function(current_df, previous_df, index_cols, compare_values = FALSE){
+  # compare_values is FALSE by default because it is computationally intensive
+  # returns a list containing: 
+  # (1) a colname comparison;
+  # (2) an index comparison;
+  # (3) IFF all column names, indices, and dims match, a values summary
+  
+  current_df <- as.data.frame(current_df)
+  previous_df <- as.data.frame(previous_df)
+  
+  # compare dimensions
+  dim_comparison <- list(
+    nrows_comparison = list(current = nrow(current_df), previous = nrow(previous_df)),
+    ncols_comparison = list(current = ncol(current_df), previous = ncol(previous_df))
+  )
+  nrows_match <- dim_comparison$nrows_comparison$current == dim_comparison$nrows_comparison$previous
+  ncols_match <- dim_comparison$ncols_comparison$current == dim_comparison$ncols_comparison$previous
+  
+  dim_comparison$nrows_comparison$match <- nrows_match
+  dim_comparison$ncols_comparison$match <- ncols_match
+  
+  # compare column names
+  colname_comparison <- dfc.setdiff_plus(names(current_df), names(previous_df))
+  colnames_match <- length(colname_comparison$only_in_first) == 0 & length(colname_comparison$only_in_second) == 0
+  
+  # compare indices
+  current_df$index <- dfc.get_concatenated_ids(current_df, index_cols)
+  previous_df$index <- dfc.get_concatenated_ids(previous_df, index_cols)
+  index_comparison <- dfc.setdiff_plus(current_df$index, previous_df$index)
+  indices_match <- length(index_comparison$only_in_first) == 0 & length(index_comparison$only_in_second) == 0
+  
+  # if everything else matches, and a values comparison was requested, go ahead and do a values comparison
+  if(nrows_match & ncols_match & colnames_match & indices_match & compare_values){
+    current_df <- current_df[order(current_df$index), ]
+    previous_df <- previous_df[order(previous_df$index), ]
+    values_comparison <- dfc.compare_df_values(current_df, previous_df, "index")
+    values_comparison$values_match <- all(values_comparison$col_summary$num_unmatched == 0)
```

Comment:
yeah, come to think of it, it's pretty weird to keep this part in the wrapper function, since it modifies the `values_comparison` object that is the output of that function.